shame_graph/src/expr/glsl_output.rs

```diff
@@ -588,74 +705,84 @@ fn is_ident(s: &str) -> bool {
 
 const ANONYMOUS_IDENTIFIER_PREFIX: &str = "_";
 
 fn single_valid_ident_for_slot(slot: &Option<String>) -> String {
     match slot {
-        Some(string) if is_ident(string) => {
-            match ident_satisfies_glsl_constraints(string) {
-                true => string.clone(),
-                false => match string.starts_with("__") {
-                    true => format!("{ANONYMOUS_IDENTIFIER_PREFIX}0{string}"),
-                    false => ANONYMOUS_IDENTIFIER_PREFIX.to_string() + string
-                },
-            }
-        }
-        _ => ANONYMOUS_IDENTIFIER_PREFIX.to_string()
+        Some(string) if is_ident(string) => match ident_satisfies_glsl_constraints(string) {
+            true => string.clone(),
+            false => match string.starts_with("__") {
+                true => format!("{ANONYMOUS_IDENTIFIER_PREFIX}0{string}"),
+                false => ANONYMOUS_IDENTIFIER_PREFIX.to_string() + string,
+            },
+        },
+        _ => ANONYMOUS_IDENTIFIER_PREFIX.to_string(),
     }
 }
 
-/// takes idents with the same name and puts numbers (1, 2...) behind the duplicates to 
-/// make them different form each other. this may result in new name collisions, therefore 
-/// the function returns the amount of edited idents. Call it repeatedly until it returns 
+/// takes idents with the same name and puts numbers (1, 2...) behind the duplicates to
+/// make them different form each other. this may result in new name collisions, therefore
+/// the function returns the amount of edited idents. Call it repeatedly until it returns
 /// 0 to make sure all the identifiers are unique.
 /// TODO: add tests, verify that it works, or rewrite it with a more elegant algorithm.
 #[allow(clippy::needless_range_loop)]
 fn deduplicate_idents_pass(idents: &mut Vec<String>) -> usize {
     let mut order = (0..idents.len()).collect::<Vec<_>>();
     order.sort_by_key(|i| &idents[*i]);
     let mut changed_counter = 0;
-    
+
     for i in 0..order.len() {
         let ai = order[i];
-        
-        for j in i+1..order.len() {
+
+        for j in i + 1..order.len() {
             let bi = order[j];
 
             let a = &idents[ai];
             let b = &idents[bi];
-            
+
             if a == b {
-                let delta = j-i;
+                let delta = j - i;
                 let b_mut = &mut idents[bi];
                 *b_mut = format!("{}{}", b_mut, delta);
                 if !ident_satisfies_glsl_constraints(b_mut) {
                     *b_mut = format!("{}{}", b_mut, 0); //append another '0' if the ident became invalid glsl, (e.g. if 'vec' became 'vec2', make it 'vec20' instead)
                 }
                 changed_counter += 1;
+            } else {
+                break;
             }
-            else {break}
         }
     }
 
     changed_counter
 }
 
-fn valid_idents_for_pool(idents: &Vec<Option<String>>) -> Vec<String> {
+fn valid_idents_for_pool(idents: &[Option<String>]) -> Vec<String> {
```

Comment:
`Vec<T>` implements `AsRef<[T]>`. `[T; N]` does also. Accepting a slice as input gives the caller more flexibility.